File: problem-solving/lc-1269-hard-number-of-ways-to-stay-in-the-same-place-after-some-steps/2-bottom-up.go
/*
https://leetcode.com/problems/number-of-ways-to-stay-in-the-same-place-after-some-steps

Complexity
----------
Time: O(S*min(S, N))
Space: O(S*min(S, N))

S = Steps
N = arrLen
*/

func min(x, y int) int {
    if x < y {
        return x
    }
    return y
}

func numWays(steps int, arrLen int) int {
    M := 1000000007
    var dp [][]int
    maxPos := min(steps, arrLen)
    for i := 0; i < steps + 1; i++ {
        dp = append(dp, make([]int, maxPos))
    }
    dp[0][0] = 1
    for s := 1; s <= steps; s++ {
        for i := 0; i < maxPos; i++ {
            dp[s][i] = dp[s - 1][i]
            if i > 0 {
                dp[s][i] = (dp[s][i] + dp[s - 1][i - 1]) % M
            }
            if i < maxPos - 1 {
                dp[s][i] = (dp[s][i] + dp[s - 1][i + 1]) % M
            }
        }
    }
    return dp[steps][0]
}

File: problem-solving/lc-1269-hard-number-of-ways-to-stay-in-the-same-place-after-some-steps/2-bottom-up.py
"""
https://leetcode.com/problems/number-of-ways-to-stay-in-the-same-place-after-some-steps

Complexity
----------
Time: O(S*min(S, N))
Space: O(S*min(S, N))

S = Steps
N = arrLen
"""

class Solution:
    def numWays(self, steps: int, arrLen: int) -> int:
        M = 1000000007
        maxPos = min(steps, arrLen)
        dp = [[0] * maxPos for _ in range(steps + 1)]
        dp[0][0] = 1
        for s in range(1, steps + 1):
            for idx in range(maxPos):
                dp[s][idx] = dp[s - 1][idx] % M
                if idx > 0:
                    dp[s][idx] = (dp[s][idx] + dp[s - 1][idx - 1]) % M
                if idx < maxPos - 1:
                    dp[s][idx] = (dp[s][idx] + dp[s - 1][idx + 1]) % M
        return dp[steps][0]
